src/PubSubClient.cpp

```diff
@@ -537,20 +538,20 @@ uint8_t PubSubClient::buildHeader(uint8_t header, uint8_t* buf, size_t length) {
     } while (len > 0 && hdrLen < MQTT_MAX_HEADER_SIZE - 1);
 
     if (len > 0) {
-        DEBUG_PSC_PRINTF("length too big %zu, left %zu, should be 0\r\n", length, len);
+        ERROR_PSC_PRINTF_P("buildHeader() length too big %zu, left %zu\n", length, len);
+        hdrLen = 0;
     }
 
     buf[MQTT_MAX_HEADER_SIZE - 1 - hdrLen] = header;
-    for (uint8_t i = 0; i < hdrLen; i++) {
-        buf[MQTT_MAX_HEADER_SIZE - hdrLen + i] = hdrBuf[i];
-    }
+    memcpy(buf + MQTT_MAX_HEADER_SIZE - hdrLen, hdrBuf, hdrLen);
     return hdrLen + 1;  // Full header size is variable length bit plus the 1-byte fixed header
 }
 
 bool PubSubClient::write(uint8_t header, uint8_t* buf, size_t length) {
     bool result = true;
     size_t rc;
     uint8_t hdrLen = buildHeader(header, buf, length);
+    if (hdrLen == 0) return false;  // exit here in case of header generation failure
```

Comment:
And here it is clear why an error state should return 0 when calling `buildHeader`